### ISSUE TITLE
Unable to open connection

### ISSUE THREAD

**Issue by mlh758** [go-goracle/goracle#66]
I feel like this should be obvious, but I'm running out of ideas on how to open a connection to my database and get started. The database is running on a remote server.

I tried using the `goracle.ConnectionParams` object like this:
```
conn := goracle.ConnectionParams{
    Username:    "username",
    Password:     "password",
    SID:               "hostname:1521/sid",
    MaxSessions: 10}
```

I tried using
`user/passw@hostname:port/sid`

I tried using
`oracle://user:passw@hostname:port/sid`

I also tried copying the connection string right out of the description in Oracle SQL Developer:
`user:password@//hostname:port/sid`

and prefixing that with `oracle://`

I have to be missing something obvious here. Normally in Rails it's something in a config file like:
```
database: '(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=tcp)(HOST=hostname)(PORT=1521)))(CONNECT_DATA=(SID=sid)))'
username: user
password: pass
```

but I can't figure out how to do this in goracle.

**Comment by mlh758** [comment]
Okay:

`user/password@(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=tcp)(HOST=hostname)(PORT=1521)))(CONNECT_DATA=(SID=sid)))`

Could an example like this go in the docs or readme? Or maybe I'm just the only person who struggled that hard with a connection string...

**Comment by mlh758** [comment]
Reopening just because I'm wondering if would be worth making the `ConnectionParams` struct handle this in `String()`?

**Comment by tgulacsi** [comment]
Yes, the "sid" part is tricky: the short name works only if tnsnames.ora is there (where TNS_ADMIN points); the host:port iff the listener listens there and nothing fancy is configured.

TL;DR; the SID can be anything the DBA configured - exactly what tnsping returns, or you can read in tnsnames.ora.

ConnectionParams holds only the username, password and the sid part, with some driver-specific params. It does not know anything about Oracle's SERVICE_NAME or SID, but you can set `(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=tcp)(HOST=hostname)(PORT=1521)))(CONNECT_DATA=(SID=sid)))` to its SID, and it will work.

**Comment by mlh758** [comment]
Mind if I PR that into the Readme just to make a more complex example of setting up a connection clear?

**Comment by tgulacsi** [comment]
No, I'll be very glad and accept such a PR!

**Comment by cjbj** [comment]
SIDs were obsoleted by service names many Oracle releases ago (circa Oracle 8).   Obviously the Oracle listener needs to be listening for a service name, but  the tnsnames.ora syntax `(SID=abc)` was effectively replaced by `(SERVICE_NAME=xyx)`.  Any doc should mention service names first, IMHO.

**Comment by tgulacsi** [comment]
I'm not a DBA - can you point me a good short introduction to the service description that tnsnames.ora contains and tnsping returns?

**Comment by cjbj** [comment]
Check out https://oracle.github.io/node-oracledb/doc/api.html#connectionstrings
